### PR TITLE
Add framework for JS version resolver

### PR DESCRIPTION
**Pull request by greysteil**
Plan here is to allow us to resolve JS peer dependencies before determining whether we want to do an update.

### CODE REVIEW

**Comment by greysteil** [comment]
@feelepxyz - this is getting close to good enough to hand over now. Note that we're seeing the same issue with Yarn as with `npm` - if there's a lockfile (so we don't fetch manifests) we get problems. Maybe easier to work around with using yarn internals?

File: helpers/yarn/lib/peer-dependency-checker.js [hunk 1]
```diff
@@ -46,7 +46,7 @@ function optionalRequirement(requirements) {
   );
 }
 
-function install_args_with_version(depName, desiredVersion, requirements) {
+function installArgsWithVersion(depName, desiredVersion, requirements) {
```
Comment (greysteil) [inline helpers/yarn/lib/peer-dependency-checker.js]: 🙈

File: lib/dependabot/update_checkers/java_script/npm_and_yarn/version_resolver.rb [hunk 2]
```diff
@@ -63,6 +63,15 @@ def run_update_check
           end
 
           def run_yarn_checker(path:)
+            # Top level dependecies are required in the peer dep checker
+            # to fetch the manifests for all top level deps which may contain
+            # "peerDependency" requirements
+            top_level_dependencies = FileParsers::JavaScript::NpmAndYarn.new(
```
Comment (greysteil) [inline lib/dependabot/update_checkers/java_script/npm_and_yarn/version_resolver.rb]: Split this out into a separate (private) method

File: helpers/yarn/lib/peer-dependency-checker.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+async function checkPeerDependencies(
+  directory,
+  depName,
+  desiredVersion,
+  requirements,
+  topLevelDependencies
+) {
+  for (let req of requirements) {
```
Comment (feelepxyz) [inline helpers/yarn/lib/peer-dependency-checker.js]: @greysteil what is going on here? We don't seem to do this for npm, just get the first requirement with a `source` when figuring out git version.

Do we want to do this for npm? So we get the directory based on the requirement file (for the installer): `directory = path.join(directory, path.dirname(requirement.file))`

Comment (feelepxyz) [reply helpers/yarn/lib/peer-dependency-checker.js]: ah wait that's not even what we do for yarn, is this to deal with workspaces that have a root lockfile?

Comment (greysteil) [reply helpers/yarn/lib/peer-dependency-checker.js]: Yeah, exactly that. I'm not sure it's the best way to deal with it - ages since I wrote it :-/

**Comment by greysteil** [comment]
OK, I'm pretty confident this won't break anything now - if things error the resolver will just return the latest version, as it would have before. I'm going to merge and we can iterate on this in a new PR.

**Pull request merged**